packages/itwinui-react/src/utils/hooks/useWarningLogger.ts

```diff
@@ -0,0 +1,69 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as React from 'react';
+import { isUnitTest } from '../functions/dev.js';
+
+const _React = React as any;
+const ReactInternals =
+  _React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
+
+/**
+ * Returns a function that can be used to log one-time warnings in dev environments.
+ *
+ * **Note**: The actual log call should be wrapped in a check against `process.env.NODE_ENV === 'development'`
+ * to ensure that it is removed from the production build output (by SWC).
+ * Read more about the [`NODE_ENV` convention](https://nodejs.org/en/learn/getting-started/nodejs-the-difference-between-development-and-production).
+ *
+ * @example
+ * const logWarning = useWarningLogger();
+ *
+ * if (process.env.NODE_ENV === 'development') {
+ *   logWarning("please don't use this")
+ * }
+ */
+export const useWarningLogger =
+  process.env.NODE_ENV === 'development' && !isUnitTest
+    ? function () {
+        const loggedRef = React.useRef(false);
+        const timeoutRef = React.useRef<number | null>(null);
+
+        // https://stackoverflow.com/a/71685253
+        const stack =
+          ReactInternals?.ReactDebugCurrentFrame?.getCurrentStack?.();
+
+        // Second line in the stack is the component name.
+        const componentName = stack?.trim().split('\n')[1]?.trim();
+
+        const prefix = componentName
+          ? `Warning (${componentName}):`
```

Comment:
One of the core maintainers of React [said](https://redirect.github.com/reactjs/react.dev/issues/3896#issuecomment-914036200):

> If you use `console.error` and the user has React DevTools then they should already see the component stack in development automatically. That’s the intended solution.

It would be nice to not rely on these hacks if it works out-of-the-box. However, the component stack was not showing up for me. It was identical with or without React Dev Tools.

![](https://github.com/user-attachments/assets/3befa94b-fb87-41e0-b8f9-1b667fb30881)

Can someone else also confirm this?

Reply:
Maybe it used to work but no longer does, in which case we can't rely on it.

I'm ok with this workaround for now. It's worth noting that it only works for React 18. We'll need to investigate how to make it work in React 19 when we add support for it.